### PR TITLE
Implement string.concat() and bytes.concat()

### PR DESCRIPTION
**Pull request by seanyoung**
This is what solc implements. Solang has implement a + b for string concatenation, which this PR removes.

Fixes: https://github.com/hyperledger/solang/issues/1558

### CODE REVIEW

**Comment by LucasSte** [comment]
> This is what solc implements. Solang has implement a + b for string concatenation, which this PR removes.
> 
> Fixes: #1558

Removing `a+b` is a breaking change and render existing contracts incompatible. Shouldn't we first warn that `a+b` is deprecated and then remove it in a later release?

File: stdlib/stdlib.c [hunk 1]
```diff
@@ -52,6 +52,8 @@ void __memcpy(void *_dest, const void *_src, uint32_t length)
     {
         *dest++ = *src++;
     }
+
+    return dest;
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply stdlib/stdlib.c:56]: It makes concatenation easier in emit; the code can simply use the return value for the destination of the next part that needs to be concatenated.

File: tests/polkadot_tests/strings.rs [hunk 2]
```diff
@@ -144,7 +144,12 @@ fn string_concat() {
         r#"
         contract foo {
             function test() public {
-                assert(hex"41424344" == "AB" + "CD");
+                assert(hex"41424344" == bytes.concat("AB", "CD", bytes(string.concat())));
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply tests/polkadot_tests/strings.rs:147]: There is a test with three arguments in the `string_abi_decode()` test. I've added some more tests.

File: src/codegen/events/polkadot.rs [hunk 3]
```diff
@@ -124,13 +124,23 @@ impl EventEmitter for PolkadotEventEmitter<'_> {
...
-                left: prefix,
-                right: value,
+                size: Expression::NumberLiteral {
+                    loc,
+                    ty: Type::Uint(32),
+                    value: first_prefix.len().into(),
+                }
+                .into(),
+                initializer: Some(first_prefix),
+            };
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply src/codegen/events/polkadot.rs:138]: Fixed

**Comment by seanyoung** [comment]
> Removing `a+b` is a breaking change and renders existing contracts incompatible. Shouldn't we first warn that `a+b` is deprecated and then remove it in a later release?

`a+b` for strings was always a non-standard extension. It is not used much; for example, in the [solana program examples](https://github.com/solana-developers/program-examples) it is not used (I've also just tested that).

As such I don't think there is need for deprecation and eventual removal. I've added an entry to `CHANGELOG.md`.

**Review by xermicus — APPROVED** [review]
LGTM, I just see some potential to optimize the concatenation into a no-op when one operand is an empty constant:

```solidity
function f(string a) public returns (string) {
    return string.concat(a, "");
}
```

```
block0: # entry
        ty:string %a = (arg #0)
        return (builtin Concat ((arg #0), (alloc string uint32 0 "")))
```

**Comment by seanyoung** [comment]
> LGTM, I just see some potential to optimize the concatenation into a no-op when one operand is an empty constant:
> 
> ```solidity
> function f(string a) public returns (string) {
>     return string.concat(a, "");
> }
> ```
> 
> ```
> block0: # entry
>         ty:string %a = (arg #0)
>         return (builtin Concat ((arg #0), (alloc string uint32 0 "")))
> ```

Nice, good catch. I've fixed this and added a test case.

File: stdlib/stdlib.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (LucasSte) [reply stdlib/stdlib.c:56]: Why is it easier? Before you only need to construct a function call inside a for loop without changing the first argument, now you need a return that is exactly the same value as the argument.

[the remark this answers is not in the document]
Comment (seanyoung) [reply stdlib/stdlib.c:56]: It does not return the same value as its argument. It returns the dest + len

**Comment by LucasSte** [comment]
> > Removing `a+b` is a breaking change and renders existing contracts incompatible. Shouldn't we first warn that `a+b` is deprecated and then remove it in a later release?
> 
> It is not used much; 

We don't know if it is not used much. There is no way to prove it. The solana examples is a small set of what people use Solang for. Polkadot also support this feature.

Instead of deprecating then, what do you think of suggesting people to use `string.concat` when erroring for `string + string`?

File: src/codegen/constant_folding.rs [hunk 4]
```diff
@@ -1924,51 +1952,89 @@ fn string_compare(
...
+            if let Some(Expression::AllocDynamicBytes {
+                size,
+                initializer: Some(init),
+                ..
+            }) = &mut last
+            {
+                if let Expression::NumberLiteral { value, .. } = size.as_mut() {
+                    *value += bs.len();
+                } else {
+                    unreachable!();
```
Comment (LucasSte) [inline src/codegen/constant_folding.rs:1986]:
```suggestion
                let Expression::NumberLiteral { value, .. } = size.as_mut() else {
                    unreachable!();
                 }
                 *value += bs.len();
```

**Review by LucasSte — APPROVED** [review]

**Comment by seanyoung** [comment]
> Instead of deprecating then, what do you think of suggesting people to use `string.concat` when erroring for `string + string`?

I've added a diagnostic for this. I think it's useful in general for folks when they may try to concatenate two strings.

**Pull request merged**